pkg/pillar/kubeapi/storageutils.go

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//go:build kubevirt
+
+package kubeapi
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"syscall"
+
+	"github.com/lf-edge/eve/pkg/pillar/pubsub"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"golang.org/x/sys/unix"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	longhornDevPath = "/dev/longhorn"
+)
+
+func getMajorMinorStr(stat syscall.Stat_t) string {
+	major := unix.Major(stat.Rdev)
+	minor := unix.Minor(stat.Rdev)
+	return fmt.Sprintf("%d:%d", major, minor)
+}
+
+func isLonghornVolAttachedToThisNode(pv string) bool {
+	if _, err := os.Stat(longhornDevPath + "/" + pv); err != nil {
+		return false
+	}
+	return true
+}
+func deviceExists(dev string) bool {
+	if _, err := os.Stat("/dev/" + dev); err != nil {
+		return false
+	}
+	return true
+}
+
+// CleanupDetachedDiskMetrics loops over existing DiskMetric
+// objects and unpublishes them if the device no longer exists.
+// Clustered volumes are expected to not be static, they will
+// move between nodes.
+func CleanupDetachedDiskMetrics(pubDiskMetric pubsub.Publication, pvcToPvMap map[string]string) {
+	existingMetrics := pubDiskMetric.GetAll()
+
+	for id, metric := range existingMetrics {
+		if id == "" {
+			continue
+		}
+
+		dm, ok := metric.(types.DiskMetric)
+		if ok && dm.IsDir {
+			continue
+		}
+
+		if strings.Contains(id, "pvc-") {
+			pvName, ok := pvcToPvMap[id]
+			// Could be PVC deleted or just not attached locally
+			if !ok || !isLonghornVolAttachedToThisNode(pvName) {
+				pubDiskMetric.Unpublish(id)
+			}
+		} else {
+			// Look for sdX devices which used to exist
+			// These would have been the block device
+			// which shared major:minor with the longhorn device
+			if !deviceExists(id) {
+				pubDiskMetric.Unpublish(id)
+			}
+		}
+	}
+}
+
+// LonghornGetMajorMinorMaps builds two maps between
+// device major:minor -> kube-pv-name/lh-volume-name
+// and kube-pv-name/lh-volume-name -> maj:min to
+// help callers find a PV/PVC in /proc/diskstats
+// which only shows the sdX path.
+func LonghornGetMajorMinorMaps() (map[string]string, map[string]string, error) {
+	lhMajMinToNameMap := make(map[string]string) // maj:min -> kube-pv-name/lh-volume-name
+	lhNameToMajMinMap := make(map[string]string) // kube-pv-name/lh-volume-name -> maj:min
+
+	if _, err := os.Stat(longhornDevPath); err != nil {
+		return lhMajMinToNameMap, lhNameToMajMinMap, fmt.Errorf("longhorn dev path missing")
+	}
+
+	lhPvcList, err := os.ReadDir(longhornDevPath)
+	if err != nil {
+		return lhMajMinToNameMap, lhNameToMajMinMap, fmt.Errorf("unable to read longhorn devs")
```

Comment:
```suggestion
		return lhMajMinToNameMap, lhNameToMajMinMap, fmt.Errorf("unable to read longhorn devs: %w", err)
```
then I think the check if `longhornDevPath` even exists is not necessary.

Reply:
implemented in latest push